File: 0-999/0-99/15.cpp
#include <bits/stdc++.h>

using namespace std;

int main(){





class Solution {
public:
  vector<vector<int>> threeSum(vector<int>& nums) {
    sort(nums.begin(), nums.end(), less<int>());
    vector<vector<int>> res;
    for (int i = 0; i < nums.size() - 2; i++) {
      int p1 = i + 1;
      int p2 = nums.size() - 1;
      while (p1 < p2) {
        int n = nums[p1] + nums[p2] + nums[i];
        if (n == 0) {
          if (p1 == i + 1 || p2 == nums.size() - 1) {
                res.push_back({nums[i], nums[p1], nums[p2]});

            }
        }
        if (n <= 0) {
          p1++;
        } else {
          p2--;
        }
      }
    }
    return res;
  }
};










  return 0;
}

File: 0-999/0-99/15.ts
// hyunseo's solution

function threeSum(nums: number[]): number[][] {
  let arr: number[][] = [];
  for(let i =0; i<nums.length; i++){
      if(twoSum(nums,nums[i], i, arr).length !== 0){
          let ans: number[] = twoSum(nums,nums[i], i, arr);
          arr.push([nums[i],ans[0], ans[1]]);
          i--
      }
  }
  return arr;
};

function twoSum(nums: number[], nein: number, start: number, arr: number[][]): number[]{
  let tar: number = -nein;
  for(let i=0; i<nums.length; i++){
      let x: number = tar-nums[i];
      let y: number = nums.indexOf(x);
      let dup: boolean = false;
      if(i!==y && y!==-1 && y!== start && i!== start) {

          if(nums[i] === nums[y] && nums[y] === nein){
              for(let j=0; j<arr.length;j++){
                  if(
                      arr[j][0] === nein &&
                      arr[j][1] === nein 
                  ) return [];
              }
              return [nums[i],nums[y]];
          }
          for(let j=0; j<arr.length; j++){
              if(
                  arr[j].includes(nums[i]) && 
                  arr[j].includes(nums[y]) &&
                  arr[j].includes(nein) 
              )
              {
                  dup = true;
              }
          }
          if(!dup) return [nums[i],nums[y]];
      }
  }
  return [];
}



function BruteForceThree(nums: number[]) {
  let res: number[][] = [];
  nums.sort();
  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      for (let k = j + 1; k < nums.length; k++) {
        if (nums[i] + nums[j] + nums[k] == 0) {
          for (let x = 0; x < res.length; x++) {
            res.push([nums[i], nums[j], nums[k]]);
            if (nums[i] == res[x][0] && nums[j] == res[x][1] && nums[k] == res[x][2]) {
              res.pop();
            }
          }
        }
      }
    }
  }
}